File: Notebook examples/stabl_knockoff_experiments/official_plsko_bridge.py
"""Python wrapper around the authors' official PLSKO R package.

The initial Python port supplied with this project is intentionally not used in
this wrapper.  PLSKO includes several implementation details and tuning rules,
so scientific comparisons should use the authors' package unless an independent
Python port has been validated by simulation against that package.
"""

from __future__ import annotations

import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

import numpy as np
import pandas as pd
from numpy.typing import ArrayLike, NDArray


@dataclass
class OfficialPLSKOSampler:
    X: ArrayLike
    threshold_abs: float | None = None
    threshold_q: float | None = 0.8
    ncomp: int | None = None
    sparsity: float = 1.0
    random_state: int = 0
    rscript: str = "Rscript"
    bridge_script: str | Path | None = None
    timeout_seconds: int = 7200

    def _bridge_path(self) -> Path:
        if self.bridge_script is not None:
            return Path(self.bridge_script).expanduser().resolve()
        return Path(__file__).with_name("official_plsko_bridge.R")

    def _validate_environment(self) -> None:
        if shutil.which(self.rscript) is None:
            raise RuntimeError(
                f"Could not find {self.rscript!r}. Install R or provide the full Rscript path."
            )
        bridge = self._bridge_path()
        if not bridge.exists():
            raise FileNotFoundError(f"PLSKO bridge script not found: {bridge}")

    def sample_knockoffs(self) -> NDArray[np.float64]:
        self._validate_environment()
        if isinstance(self.X, pd.DataFrame):
            X_df = self.X.copy()
        else:
            X_arr = np.asarray(self.X, dtype=float)
            if X_arr.ndim != 2:
                raise ValueError("X must be a two dimensional matrix")
            X_df = pd.DataFrame(
                X_arr,
                index=[f"sample_{i}" for i in range(X_arr.shape[0])],
                columns=[f"feature_{j}" for j in range(X_arr.shape[1])],
            )
        if not np.isfinite(X_df.to_numpy(dtype=float)).all():
            raise ValueError("OfficialPLSKOSampler currently requires complete finite X")
        if not 0.0 < self.sparsity <= 1.0:
            raise ValueError("sparsity must lie in (0, 1]")

        with tempfile.TemporaryDirectory(prefix="stabl_plsko_") as tmp:
            tmp_path = Path(tmp)
            input_csv = tmp_path / "X.csv"
            output_csv = tmp_path / "X_knockoff.csv"
            X_df.to_csv(input_csv)

            command = [
                self.rscript,
                str(self._bridge_path()),
                "--input",
                str(input_csv),
                "--output",
                str(output_csv),
                "--seed",
                str(int(self.random_state)),
                "--sparsity",
                str(float(self.sparsity)),
            ]
            if self.threshold_abs is not None:
                command += ["--threshold-abs", str(float(self.threshold_abs))]
            if self.threshold_q is not None:
                command += ["--threshold-q", str(float(self.threshold_q))]
            if self.ncomp is not None:
                command += ["--ncomp", str(int(self.ncomp))]

            completed = subprocess.run(
                command,
                text=True,
                capture_output=True,
                timeout=self.timeout_seconds,
                check=False,
            )
            if completed.returncode != 0:
                raise RuntimeError(
                    "Official PLSKO R process failed.\n"
                    f"Command: {' '.join(command)}\n"
                    f"stdout:\n{completed.stdout}\n"
                    f"stderr:\n{completed.stderr}"
                )
            if not output_csv.exists():
                raise RuntimeError("PLSKO completed but did not create the output CSV")

            Xk_df = pd.read_csv(output_csv, index_col=0)
            Xk_df = Xk_df.loc[X_df.index, X_df.columns]
            Xk = Xk_df.to_numpy(dtype=float)
            if Xk.shape != X_df.shape:
                raise RuntimeError(f"PLSKO returned {Xk.shape}, expected {X_df.shape}")
            if not np.isfinite(Xk).all():
                raise RuntimeError("PLSKO returned NaN or infinite values")
            self.X_artificial_ = Xk
            return Xk


def check_official_plsko_installation(
    rscript: str = "Rscript", bridge_script: str | Path | None = None
) -> tuple[bool, str]:
    bridge = (
        Path(bridge_script).expanduser().resolve()
        if bridge_script is not None
        else Path(__file__).with_name("official_plsko_bridge.R")
    )
    if shutil.which(rscript) is None:
        return False, f"Could not find {rscript}"
    command = [
        rscript,
        "-e",
        "quit(status=ifelse(requireNamespace('PLSKO', quietly=TRUE), 0, 2))",
    ]
    result = subprocess.run(command, text=True, capture_output=True, check=False)
    if result.returncode == 2:
        return False, "R is available, but package PLSKO is not installed"
    if result.returncode != 0:
        return False, result.stderr.strip() or "Unknown R error"
    if not bridge.exists():
        return False, f"Bridge script is missing: {bridge}"
    return True, "Official PLSKO R package and bridge are available"

File: Notebook examples/stabl_knockoff_experiments/official_plsko_bridge.R
#!/usr/bin/env Rscript

# Command line bridge to the authors' official PLSKO R package.
#
# Example:
#   Rscript official_plsko_bridge.R \
#     --input X.csv --output X_knockoff.csv --seed 1 \
#     --threshold-q 0.8 --ncomp 5 --sparsity 1

args <- commandArgs(trailingOnly = TRUE)

read_flag <- function(name, default = NULL) {
  pos <- which(args == name)
  if (length(pos) == 0) return(default)
  if (pos[1] == length(args)) stop(sprintf("Missing value after %s", name))
  args[pos[1] + 1]
}

input_path <- read_flag("--input")
output_path <- read_flag("--output")
seed <- as.integer(read_flag("--seed", "1"))
threshold_abs_raw <- read_flag("--threshold-abs", NA_character_)
threshold_q_raw <- read_flag("--threshold-q", "0.8")
ncomp_raw <- read_flag("--ncomp", NA_character_)
sparsity_raw <- read_flag("--sparsity", "1")

if (is.null(input_path) || is.null(output_path)) {
  stop("Both --input and --output are required")
}
if (!requireNamespace("PLSKO", quietly = TRUE)) {
  stop(
    paste0(
      "The PLSKO package is not installed. Run in R:\n",
      "  install.packages('devtools')\n",
      "  devtools::install_github('guannan-yang/PLSKO/PLSKO', ",
      "quiet = TRUE, upgrade = 'never')"
    )
  )
}

X_df <- read.csv(input_path, row.names = 1, check.names = FALSE)
X <- as.matrix(X_df)
storage.mode(X) <- "double"
if (any(!is.finite(X))) {
  stop("Official PLSKO bridge currently requires a complete finite matrix")
}

set.seed(seed)
call_args <- list(X = X)
if (!is.na(threshold_abs_raw)) call_args$threshold.abs <- as.numeric(threshold_abs_raw)
if (!is.na(threshold_q_raw)) call_args$threshold.q <- as.numeric(threshold_q_raw)
if (!is.na(ncomp_raw)) call_args$ncomp <- as.integer(ncomp_raw)
if (!is.na(sparsity_raw)) call_args$sparsity <- as.numeric(sparsity_raw)

Xk <- do.call(PLSKO::plsko, call_args)
Xk <- as.matrix(Xk)
if (!all(dim(Xk) == dim(X))) {
  stop(sprintf(
    "PLSKO returned shape %s but expected %s",
    paste(dim(Xk), collapse = " x "),
    paste(dim(X), collapse = " x ")
  ))
}
rownames(Xk) <- rownames(X)
colnames(Xk) <- colnames(X)
write.csv(Xk, output_path, row.names = TRUE, quote = FALSE)
